### ISSUE TITLE
Autocomplete/suggestions take 9s+ [0.31]

### ISSUE THREAD

**Issue by greggilbert** [castwide/vscode-solargraph#99]
I think this has been an issue for me for awhile, but I've noticed it more and more.

Basically, suggestions consistently take around 9 seconds.  I've tried `Stri(Ctrl+C)`, `String.new.(Ctrl+C)`, `Enum(Ctrl+C)`, and it's always the same. Developer Tools doesn't show any errors. I've got a `.solargraph.yml` which excludes `.git` and a number of other folders.

These are the only changes in my `settings.json`:

```
    "ruby.interpreter.commandPath": "/Users/Greg/.rvm/rubies/ruby-2.5.1/bin/ruby",
    "solargraph.commandPath": "/Users/Greg/.rvm/gems/ruby-2.5.1/bin/solargraph",
```

Seems like it's good for everyone else, so I'm not sure what's going on. Any ideas?

**Comment by castwide** [comment]
How large is your workspace? E.g., number of files, total lines of code, and included gems

How big is the file you're editing? In my experience, performance can degrade with files above ~2000 lines.

Is the workspace on a local drive or a network share?

You can also try to setting `solargraph.logLevel` to `info` for more information about what it's doing.

**Comment by greggilbert** [comment]
I've got 15 folders in my workspace, but only three of them are Ruby. The Ruby ones are, I'd say, total less than 10k LOC. Number of gems in each varies, but the Gemfiles are average 125 lines long?

The workspace is a local drive (SSD).

"performance can degrade with files above ~2000 lines." - does that mean that one file with 2k lines will cause issues?

**Comment by castwide** [comment]
> I've got 15 folders in my workspace, but only three of them are Ruby. The Ruby ones are, I'd say, total less than 10k LOC. Number of gems in each varies, but the Gemfiles are average 125 lines long?

That doesn't sound big enough to cause so much lag. It might have a significant impact on initialization time, but not so much in completion suggestions.

> "performance can degrade with files above ~2000 lines." - does that mean that one file with 2k lines will cause issues?

Yes, specifically if it's the one you're currently editing. The workspace as a whole should be able to handle tens of thousands of lines.

**Comment by greggilbert** [comment]
The file I'm using is 24 lines long and just has one class in it.

I turned on logging to `debug`, and it looks like it's indexing multiple times. Every time I type another letter, it reruns everything: `Indexing workspace files in /PATH`. Is it supposed to do that?

**Comment by castwide** [comment]
No, that's definitely unusual. `Cataloging` is a frequent operation, but `indexing` the workspace shouldn't happen every time you type. It usually happens when a workspace file gets created or renamed. Do you also see a lot of messages for `workspace/didChangeWatchedFiles`?

Are you using the most current versions of the extension (0.19.1) and the gem (0.31.1)?

**Comment by greggilbert** [comment]
I see `console.ts:134 [Extension Host] [INFO] Server received textDocument/didChange`. Not the same, though.

And yeah, this is using the latest extension and gem.

**Comment by castwide** [comment]
The `textDocument/didChange` messages are normal. The `Indexing workspace files` messages shouldn't happen anywhere near that often, especially for a change to a file that's already open.

Are you working in a multi-root workspace?

**Comment by YoannNovaxis** [comment]
I got the same problem, loading time was very long for suggestions and autocomplete. I just deleted the .solargraph.yml file and it worked...

**Comment by YoannNovaxis** [comment]
@castwide I did not have the problem after generating default .solargraph.yml file.

When I had the problem the file looked like 

```
include:
  - ./**/*.rb
exclude:
  - spec/**/*
```

The default file with command solargraph config look like

```
---
include:
- "**/*.rb"
exclude:
- spec/**/*
- test/**/*
- vendor/**/*
- ".bundle/**/*"
require: []
domains: []
reporters:
- rubocop
- require_not_found
plugins: []
require_paths: []
max_files: 5000
```

It seems that some options correct the problem.

**Comment by castwide** [comment]
My guess, the additional `exclude` rules fixed it, especially if Solargraph was trying to map files for cached gems.

**Comment by greggilbert** [comment]
@castwide: Yes, it's multi-root. When I said there were 15 folders, I meant each of those is a separate root. An edit in any file causes them all to reindex?

**Comment by castwide** [comment]
Yep, that's the problem. It appears to be specific to multi-root. Editing a file should not require a reindex of any of the folders, including the one that contains the file.

I should be able to work a fix into the next patch release.

**Comment by castwide** [comment]
The fix is published in gem v0.31.2.

**Comment by greggilbert** [comment]
@castwide It's much, _much_ better now. And I can confirm that I'm not seeing constant reindexing. Thanks!
